src/bank.rs

```diff
@@ -375,26 +416,23 @@ impl Bank {
         self.register_entry_id(&entry0.id);
         self.register_entry_id(&entry1.id);
 
-        let mut entry_count = 2;
         let mut tail = Vec::with_capacity(WINDOW_SIZE as usize);
-        let mut next = Vec::with_capacity(WINDOW_SIZE as usize);
-
-        for block in &entries.into_iter().chunks(WINDOW_SIZE as usize) {
-            tail = next;
-            next = block.collect();
-            entry_count += self.process_blocks(next.clone())?;
-        }
+        tail.push(entry0);
+        tail.push(entry1);
+        let mut tail_idx = 2;
+        let entry_count = 2 + self.process_blocks(entries, &mut tail, &mut tail_idx)?;
+
+        // check if we need to shift tail around
+        let tail = if tail.len() == WINDOW_SIZE as usize && tail_idx != 0 {
+            let mut shift = Vec::with_capacity(WINDOW_SIZE as usize);
+            shift.extend_from_slice(&tail[tail_idx..]);
+            shift.extend_from_slice(&tail[0..tail_idx]);
```

Comment:
Any way that can be implemented with higher-level `Vec` functions. Maybe something like:

```rust
let mut (left, right) = v.split_at(tail_idx);
right.extend(left);
right
```

Alternatively, if you move this code into a separate function and write a test that demonstrates your intent here, I don't need to grok the implementation - just the test.

Reply:
what you've written is equivalent, with more mallocs()

Reply:
Right, it's more complex than the simplest correct version, which is why you'd create a separate function and test for it

Reply:
which would leave someone else the opportunity to either further optimize it and see that it still meets your requirements, as well as leave space for someone to add a benchmark and possibly determine the compiler already implements your optimization.

Reply:
```
 fn test_rotate_vector() {
        let test = vec![4, 1, 2, 3];
        let etest = vec![4, 1, 2, 3];

        let (left, right) = etest.split_at(1);
        let mut right = right.to_vec();
        right.extend(left);

        assert_eq!(rotate_vector(test, 1), right);

        let now = Instant::now();
        for _ in 0..100_000 {
            let etest = vec![4, 1, 2, 3];
            let (left, right) = etest.split_at(1);
            let mut right = right.to_vec();
            right.extend(left);
        }
        eprintln!("split_at {}", duration_as_us(&now.elapsed()));

        let now = Instant::now();
        for _ in 0..100_000 {
            let test = vec![4, 1, 2, 3];
            rotate_vector(test, 1);
        }
        eprintln!("rotate {}", duration_as_us(&now.elapsed()));
    }
```
split_at 97145
rotate 70582